cirq/protocols/has_unitary.py

```diff
@@ -0,0 +1,182 @@
+
+    def _has_unitary_(self) -> bool:
+        """Determines whether the receiver has a unitary effect.
+
+        This method is used preferentially by the global `cirq.has_unitary`
+        method, because this method is much cheaper than the fallback strategies
+        such as checking `value._unitary_` (which causes a large matrix to be
+        computed).
+
+        Returns:
+            Whether or not the receiving object (`self`) has a unitary effect.
+        """
+
+
+def has_unitary(val: Any) -> bool:
+    """Determines whether the value has a unitary effect.
+
+    Determines whether `val` has a unitary effect by attempting the following
+    strategies:
+
+    1. Try to use `val.has_unitary()`.
+        Case a) Method not present or returns `NotImplemented`.
+            Inconclusive.
+        Case b) Method returns `True`.
+            Unitary.
+        Case c) Method returns `False`.
+            Not unitary.
+
+    2. Try to use `val._decompose_()`.
+        Case a) Method not present or returns `NotImplemented` or `None`.
+            Inconclusive.
+        Case b) Method returns an OP_TREE containing only unitary operations.
+            Unitary.
+        Case c) Method returns an OP_TREE containing non-unitary operations.
+            Not Unitary.
+
+    3. Try to use `val._apply_unitary_(args)`.
+        Case a) Method not present or returns `NotImplemented`.
+            Inconclusive.
+        Case b) Method returns a numpy array.
+            Unitary.
+        Case c) Method returns `None`.
+            Not unitary.
+
+    4. Try to use `val._unitary_()`.
+        Case a) Method not present or returns `NotImplemented`.
+            Continue to next strategy.
+        Case b) Method returns a numpy array.
+            Unitary.
+        Case c) Method returns `None`.
+            Not unitary.
+
+    It is assumed that, when multiple of these strategies give a conclusive
+    result, that these results will all be consistent with each other. If all
+    strategies are inconclusive, the value is classified as non-unitary.
+
+    Args:
+        The value that may or may not have a unitary effect.
+
+    Returns:
+        Whether or not `val` has a unitary effect.
+    """
+    strats = [
+        _strat_has_unitary_from_has_unitary, _strat_has_unitary_from_decompose,
+        _strat_has_unitary_from_apply_unitary, _strat_has_unitary_from_unitary
+    ]
+    for strat in strats:
+        result = strat(val)
+        if result is not None:
+            return result
+
+    # If you can't tell that it's unitary, it's not unitary.
+    return False
+
+
+def _strat_has_unitary_from_has_unitary(val: Any) -> Optional[bool]:
+    """Attempts to infer a value's unitary-ness via its _has_unitary_ method."""
+    getter = getattr(val, '_has_unitary_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is NotImplemented:
+        return None
+    return result
```

Comment:
Why is this function so convoluted? If there is no `_has_unitary_` then the flow is: `getter = None -> result = NotImplemented -> return = None` which seems roundabout.

Maybe: 
```python
if hasattr(val, '_has_unitary_'):
  result = getattr(val, '_has_unitary_')()
  if result is NotImplemented:
    return None
  return result
return None
```

The behavior would be different for classes that explicitly define a class attribute `_has_unitary_ = None` but I don't think that should ever happen (?)

Reply:
That rewrite does two attribute lookups instead of one... but yes I agree it's clearer so I'll use it.

Reply:
> That rewrite does two attribute lookups instead of one

I knew someone would point this out :) Caring about performance while using `getattr` is a bit of an oxymoron though :)